spark/src/main/java/org/apache/iceberg/spark/actions/BaseRemoveFilesSparkAction.java

```diff
@@ -0,0 +1,216 @@
+ * to determine the files that needs to be deleted. The action uses metadata tables to
+ * find the files to be deleted.
+ * Deletes are performed locally after retrieving the results from the Spark executors.
+ */
+@SuppressWarnings("UnnecessaryAnonymousClass")
+public class BaseRemoveFilesSparkAction
+    extends BaseSparkAction<RemoveFiles, RemoveFiles.Result> implements RemoveFiles {
+  private static final Logger LOG = LoggerFactory.getLogger(BaseRemoveFilesSparkAction.class);
+
+  private static final String DATA_FILE = "Data File";
+  private static final String MANIFEST = "Manifest";
+  private static final String MANIFEST_LIST = "Manifest List";
+  private static final String OTHERS = "Others";
+
+  private static final String STREAM_RESULTS = "stream-results";
+
+  // Creates an executor service that runs each task in the thread that invokes execute/submit.
+  private static final ExecutorService DEFAULT_DELETE_EXECUTOR_SERVICE = null;
+
+  private final Table table;
+  private final TableOperations ops;
+  private final Consumer<String> defaultDelete = new Consumer<String>() {
+    @Override
+    public void accept(String file) {
+      ops.io().deleteFile(file);
+    }
+  };
+
+  private Consumer<String> deleteFunc = defaultDelete;
+  private ExecutorService deleteExecutorService = DEFAULT_DELETE_EXECUTOR_SERVICE;
+
+  public BaseRemoveFilesSparkAction(SparkSession spark, Table table) {
+    super(spark);
+    this.table = table;
+    this.ops = ((HasTableOperations) table).operations();
+  }
+
+  @Override
+  protected RemoveFiles self() {
+    return this;
+  }
+
+  @Override
+  public Result execute() {
+    JobGroupInfo info = newJobGroupInfo("REMOVE-FILES", "REMOVE-FILES");
+    return withJobGroupInfo(info, this::doExecute);
+  }
+
+  private Result doExecute() {
+    boolean streamResults = PropertyUtil.propertyAsBoolean(options(), STREAM_RESULTS, false);
+    Dataset<Row> validFileDataset = buildValidFileDF(ops.current()).distinct();
+    RemoveFilesActionResult result;
+    if (streamResults) {
+      result = deleteFiles(validFileDataset.toLocalIterator());
+    } else {
+      result = deleteFiles(validFileDataset.collectAsList().iterator());
+    }
+    return result;
+  }
+
+  private Dataset<Row> appendTypeString(Dataset<Row> ds, String type) {
+    return ds.select(new Column("file_path"), functions.lit(type).as("file_type"));
+  }
+
+  private Dataset<Row> buildValidFileDF(TableMetadata metadata) {
+    Table staticTable = newStaticTable(metadata, this.table.io());
+    return appendTypeString(buildValidDataFileDF(staticTable), DATA_FILE)
+        .union(appendTypeString(buildManifestFileDF(staticTable), MANIFEST))
+        .union(appendTypeString(buildManifestListDF(staticTable), MANIFEST_LIST))
+        .union(appendTypeString(buildOtherMetadataFileDF(ops), OTHERS));
+  }
+
+  protected Dataset<Row> buildOtherMetadataFileDF(TableOperations tableOps) {
```

Comment:
`BaseAction` and `BaseSparkAction` seem to have similar methods `getOtherMetadataFilePaths` and `getOtherMetadataFilePaths` that are less comprehensive than this since this seems to look for `metadata.previousFiles` in older metadata files, and their only usage is within `BaseRemoveOrphanFilesSparkAction`; wondering if we want to clean them up/if remove orphan action would benefit from it, although it could be a separate issue/PR

Reply:
I think that method should be removed from `BaseAction` (not used and we want to get rid of the old action hierarchy completely). This logic should probably be placed into a utility class so that Flink and Spark can use it.

I think the remove orphan files should continue to load just the reachable json files and don't open previous json files. It should be safe to drop too old json files in that action.

Reply:
@karuppayya, I think we can combine the new logic with the existing one. What about creating a utility class in a separate PR?

We could do something like

```
org.apache.iceberg.util.MetadataLocationUtil
MetadataLocationUtil.manifestListLocations(snapshots) -> List<String>
MetadataLocationUtil.miscMetadataFileLocations(ops, recursively) -> Set<String>
```

Reply:
Filed https://github.com/apache/iceberg/issues/2559 to track this